[Yan Weidong (Evan)] iP

evanyan13 opened the pull request:
# Jarvis CLI Task Manager
> "Track and Manage your tasks at your CLI"

![Screenshot 2023-09-10 at 5 55 51 PM](https://github.com/nus-cs2103-AY2324S1/ip/assets/103996156/2cffba00-e01c-478d-a789-bafcb817b0d2)

**Hi Master! Jarvis is here at your service. Jarvis is a CLI-based Task Manager for:**
- Text-based
- Easy to Learn
- Fast to Use

### All you need to do is: 
1. Download it from [this link](https://github.com/evanyan13/ip/releases/tag/A-Jar)
2. Run java -jar jarvis.jar in the directory where the file is present
3. Add your tasks.
4. Let it manage your tasks for you 😁

### Features:
- [x]  Managing Tasks
- [x]  Filter Tasks based on keyword
- [ ]  Managing Deadlines (coming soon)

Here's the `main` method which can be founded in `src/main/java/jarvis/Jarvis.java`

```java
public static void main(String[] args) {
        Jarvis jarvis = new Jarvis();
        jarvis.start();

        Scanner scanner = new Scanner(System.in);
        String userInput;

        try {
            while (true) {
                userInput = scanner.nextLine();
                jarvis.respond(userInput);
            }
        } finally {
            scanner.close();
        }
    }

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
First Commit for Level 0

Commit:
Edit Chatbot name and logo

Commit:
Add Main class

evanyan13 committed:
Echo user command and exit with bye

evanyan13 committed:
Implement TaskList and Task class to store and display tasks. Add UI class to implement all output to the CLI

evanyan13 committed:
Implement Update Status Function in Jarvis class

evanyan13 committed:
Create Todo, Deadline and Event subclass of Task to keep track of specific information of each task

evanyan13 committed:
Update IO Redirection Tests

evanyan13 committed:
Add Exception Handling

evanyan13 committed:
Implement delete function to delete tasks by their index in the taskList

evanyan13 committed:
Update InvalidCommandException Message

evanyan13 committed:
Add save and load tasks function from file destination

evanyan13 committed:
Merge pull request #1 from evanyan13/branch-Level-7

Add save and load tasks function from file destination

evanyan13 committed:
Add load and save function

evanyan13 committed:
Add parsing and formatting for date and time

evanyan13 committed:
Merge pull request #2 from evanyan13/branch-Level-8

Add support for parsing and formatting date and time in multiple formats

evanyan13 committed:
Add command package to handle all the commands from user

evanyan13 committed:
Merge pull request #3 from evanyan13/add-command-package

Add command package to handle all the commands from user

evanyan13 committed:
Merge pull request #4 from evanyan13/branch-A-MoreOOP

Implement more OOP in code design and package similar functions together

evanyan13 committed:
Add task package to gather all implementation for tasks

evanyan13 committed:
Merge pull request #5 from evanyan13/add-task-package

Add task package to gather all implementation for tasks

evanyan13 committed:
Merge pull request #6 from evanyan13/branch-A-MoreOOP

Create task package

evanyan13 committed:
Merge branch 'add-gradle-support'

evanyan13 committed:
Add Gradle support

evanyan13 committed:
Update gradle wrapper class

evanyan13 committed:
Stage commit for gradlew

evanyan13 committed:
Merge pull request #7 from evanyan13/update-gradle

Update gradle support

evanyan13 committed:
Package app as a JAR file

evanyan13 committed:
Add JavaDoc for non-private class and methods

evanyan13 committed:
Merge pull request #8 from evanyan13/branch-A-Junit

Package app as a JAR file

evanyan13 committed:
Edit README

evanyan13 committed:
Edit code to adhere to Java Coding Standard

evanyan13 committed:
Add find function to find task based on keyword

evanyan13 committed:
Merge pull request #9 from evanyan13/branch-A-CodingStandard

Implement Java coding standard through the code

evanyan13 committed:
Merge branch 'master' into branch-Level-9

evanyan13 committed:
Merge pull request #10 from evanyan13/branch-Level-9

Add find function to find task based on keyword

evanyan13 committed:
Add Javadoc for FindCommand and Task class

tanyyyming commented on src/main/java/jarvis/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     * @return A Task object representing the parsed task.
+     * @throws InvalidTaskFormatException If the task format is invalid.
+     */
+    public static Task parseStringToTask(final String line) throws InvalidTaskFormatException {
```
If I read correctly, this should be the method that reads from the data file and converts the string to task? If so, would it be better to name it `decodeStringToTask` to avoid confusion with the method that parses the user input?
```suggestion
    public static Task encodeStringToTask(final String line) throws InvalidTaskFormatException {
```

tanyyyming commented on src/main/java/jarvis/Parser.java:
```diff
@@ -0,0 +1,134 @@
+     */
+    public static Task parseStringToTask(final String line) throws InvalidTaskFormatException {
+        try {
+            String[] lineSplit = line.split("\\|");
```
Is it better if the variable name `lineSplit` is replaced with `splittedLine`?

evanyan13 replied:
Good point! Thank you and I have implemented the changes

tanyyyming commented on src/main/java/jarvis/Parser.java:
```diff
@@ -0,0 +1,134 @@
+            boolean isCompleted = Integer.parseInt(lineSplit[1].trim()) == 1 ? true : false;
+            String taskDetails = lineSplit[2].trim();
+
+            switch (taskType) {
+            case "T":
+                return new Todo(taskDetails, isCompleted);
+            case "D":
+                String deadlineByString = lineSplit[3].trim();
+                LocalDateTime formattedDeadlineBy = parseDateTime(deadlineByString);
+                return new Deadline(taskDetails, formattedDeadlineBy, isCompleted);
+            case "E":
+                String fromTime = lineSplit[3].trim();
+                String toTime = lineSplit[4].trim();
+                LocalDateTime formattedFromTime = parseDateTime(fromTime);
+                LocalDateTime formattedToTime = parseDateTime(toTime);
+                return new Event(taskDetails, formattedFromTime, formattedToTime, isCompleted);
+            default:
+                throw new InvalidTaskFormatException("Unknown Task Type: " + taskType);
+            }
```
Good naming of local variables in this section!

tanyyyming commented on src/main/java/jarvis/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws InvalidIndexException {
+        String[] userInputSpilt = userInput.split(" ");
```
Same here. I don't know whether `userInputSplit` is a good name for a variable since it sounds like a verb (method).

tanyyyming commented on src/main/java/jarvis/tasks/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class Deadline extends Task {
+
+    private LocalDateTime dueDate;
```
Maybe rename to `dueDateTime` to be more consistent with the other variables such as `toDateTime` and `fromDateTime`?
```suggestion
    private LocalDateTime dueDateTime;
```

tanyyyming commented on src/main/java/jarvis/tasks/Task.java:
```diff
@@ -0,0 +1,101 @@
+     * @param keyword The keyword the user wish to search for.
+     * @return True if the task title contains the keyword, false otherwise.
+     */
+    public boolean contains(String keyword) {
+        if (title.contains(keyword)) {
+            return true;
+        } else {
+            return false;
+        }
+    }
```
Since this is a boolean method, is it better if you prefix it with keywords like `has`? Maybe something like `hasTasksWithKeyword`? Though I know that the string method `contains` also returns a boolean value, so its definitely up to you!
And by the way, the method can be simplified.
```suggestion
    public boolean hasTaskWithKeyword(String keyword) {
        return title.contains(keyword);
    }
```

tanyyyming reviewed (commented):
In general, the naming standards are followed pretty closely. Good job!

I've pointed out some minor modifications which are not serious offences of the standard. It is up to you to change it or keep it!

evanyan13 committed:
Update src/main/java/jarvis/tasks/Deadline.java

update var name dueDate -> dueDateTime

Co-authored-by: tanyyyming <113348423+tanyyyming@users.noreply.github.com>

evanyan13 committed:
Edit to adhere to coding standard

evanyan13 committed:
Merge branch 'master' of https://github.com/evanyan13/ip

evanyan13 committed:
Edit syntax errors

evanyan13 committed:
Update gradle support

evanyan13 committed:
Update README.md

evanyan13 committed:
Add CheckStyle to app

evanyan13 committed:
Merge pull request #11 from evanyan13/branch-A-CheckStyle

Add CheckStyle

evanyan13 committed:
Add CheckStyle

evanyan13 committed:
Merge pull request #12 from evanyan13/branch-A-CheckStyle

Add CheckStyle to current app

evanyan13 committed:
Add sample JUnit tests

evanyan13 committed:
Enable shadowJar via Gradle

evanyan13 committed:
Merge pull request #13 from evanyan13/branch-A-Jar

Enable shadowJar via Gradle

evanyan13 committed:
Add GUI feature to Jarvis app

evanyan13 committed:
Update gui print message functions

evanyan13 committed:
Add GUI to Jarvis app

evanyan13 committed:
Edit syntax error

evanyan13 committed:
Edit ui output format of Jarvis

evanyan13 committed:
Edit code to adhere to Java checkstyle

evanyan13 committed:
Add assertions to Main class

evanyan13 committed:
Add assertions to document important assumptions and conditions

The current code lacks assertions to clearly document critical assumptions and conditions at various points. This absence can lead to misunderstandings, make debugging harder, and potentially result in runtime errors.

To address this issue, I have added appropriate assertions using the assert feature. These assertions serve as documentation to ensure that specific conditions are met during program execution.

By doing so, we enhance code clarity, make it easier to understand and maintain, and assist in identifying problems early in development or testing.

evanyan13 committed:
Refactors to improve code quality

The code lacks separation of abstraction and readability issues.

Refactors the code to separate input validation, core logic, and response generation into distinct methods to adhere to best practices

This approach follows established coding principles, enhancing code readability and maintainability.

evanyan13 committed:
Add case-sensitive feature to find function

evanyan13 committed:
Implement Java Stream for find command

The code lacks the utilization of Java's Streams feature for collection processing.

To enhance code readability and potentially improve the efficiency of working with collections.

Incorporating Java Streams in the code to streamline collection operations, making the code more concise and potentially more efficient.

evanyan13 committed:
Merge pull request #16 from evanyan13/branch-A-Assertions

Add assertions to document important assumptions and conditions

evanyan13 committed:
Merge branch 'master' into branch-A-CodeQuality

evanyan13 committed:
Merge pull request #17 from evanyan13/branch-A-CodeQuality

Refactor to improve code quality

evanyan13 committed:
Edit syntax error

evanyan13 committed:
Set up GitHub Action for CI

The project lacks a Continuous Integration (CI) system.

To establish automated build and testing processes, ensuring code quality and reliability throughout the development lifecycle.

mplementing GitHub Actions to set up Continuous Integration (CI) for the project.

evanyan13 committed:
Edit based on CI feedback

evanyan13 committed:
Merge pull request #19 from evanyan13/branch-A-CI

Set up GitHub Action for CI

evanyan13 committed:
Edit gui to fit wrapped text

evanyan13 committed:
Improve gu

evanyan13 committed:
Correct Checkstyle errors

evanyan13 committed:
Edit gui to fit gif

evanyan13 committed:
Edit messageBox config in gui

evanyan13 committed:
Implement Sort Deadline Feature

Added a new feature to sort deadline tasks by their due dates in the TaskList class.

Modified the SortDeadlineCommand to display the sorted list to the user but not saving it to the jarvis.txt yet.

Consider future addition to allow the user to choose to save the filtered order to the storage file or not.

evanyan13 committed:
Save tasks to taskList after sort by deadline

evanyan13 committed:
Merge pull request #21 from evanyan13/branch-C-Sort

Add Sort Deadline function to sort deadlines by due date

evanyan13 committed:
Add User Guide

evanyan13 committed:
Create _config.yml

evanyan13 committed:
Add config to github page

evanyan13 committed:
Update README.md

evanyan13 committed:
Update User Guide

evanyan13 committed:
Merge branch 'master' of https://github.com/evanyan13/ip

evanyan13 committed:
Update User Guide

evanyan13 committed:
Edit gui

evanyan13 committed:
Edit syntax error

evanyan13 committed:
Merge pull request #23 from evanyan13/branch-A-BetterGui

Update a better gui

evanyan13 committed:
Update build.gradle in attempt to produce runnable jar

evanyan13 committed:
Merge branch 'master' of https://github.com/evanyan13/ip

evanyan13 committed:
Update shadowJar in build.gradle

evanyan13 committed:
Update new jar

evanyan13 committed:
Edit User Guide

evanyan13 committed:
Merge pull request #24 from evanyan13/configure-Jar

Edit User Guide

evanyan13 committed:
Update _config.yml

Add theme to github-page deployment

evanyan13 committed:
Merge pull request #25 from evanyan13/branch-github-page

Update _config.yml for github theme

evanyan13 committed:
Edit build gradle shadowJar config

evanyan13 committed:
Merge pull request #26 from evanyan13/configure-Jar

Edit build gradle shadowJar config

evanyan13 committed:
Add help function and help menue

evanyan13 committed:
Merge pull request #27 from evanyan13/branch-C-Help

Add help function and help menu

evanyan13 committed:
Config build.gradle to include dependencies for other OS

evanyan13 committed:
Merge pull request #28 from evanyan13/config-Jar-B

Config build.gradle to include dependencies for other OS

evanyan13 committed:
Customise Gui

evanyan13 committed:
Update README.md

evanyan13 committed:
Merge pull request #29 from evanyan13/branch-BetterGui

Customise Gui
